fix: No JS error when plugin is registered for non existing element

aragon999 opened the pull request:
### 1. Why is this change necessary?
Since Shopware 6.4.8.0 there is an error thrown, if one tries to register a plugin for an non-existing element:
![grafik](https://user-images.githubusercontent.com/6317761/153498906-12388934-5569-44a2-8d7b-ac47faa1d65e.png)

### 2. What does this change do, exactly?
Catch the case when the error would occur and do not trigger it. The other option would be to implement a strict mode for `_initializePluginOnElement`.

### 3. Describe each step to reproduce the issue or behaviour.
\-

### 4. Please link to the relevant issues (if any).
\-

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

aragon999 committed:
fix: No JS error when plugin is registered for non existing element

tobiasberge commented:
Hi @aragon999, thanks for your contribution. 👍 

The reasons seems to be that inside `_queryElements` https://github.com/shopware/platform/blob/trunk/src/Storefront/Resources/app/storefront/src/plugin-system/plugin.manager.js#L310 it will return an array with null `[null]` instead of an empty `NodeList[]` in case of a missing id.

The checks further down the line inside `_initializePluginOnElement` would now only be skipped when it's an id. I would actually prefer a solution inside `_queryElements` so that it always returns a `NodeList[]` somehow like `querySelectorAll`. We will revert the change which introduced `_queryElements` because of this error but we would like to re-introduce it in the future with an adjusted implementation.

aragon999 commented:
I understand, thanks.

I will adapt this PR and fix the error in a different way. This was just the easiest solution so far.

tobiasberge commented:
Hi, sorry I made a mistake. 🤯 I debugged a little further, and when there is an empty `nodeList` (when using a non existent class selector) the part with `Iterator.iterate(selector, el => {}` is actually not iterating because the `nodeList` or `HTMLCollection` is empty. But right now with the id case it suddenly iterates over an array containing one item with `null`. And this is causing the error further down. So your fix was actually doing kind of the same as when the `NodeList` would have been empty because `_initializePluginOnElement` is not running.

Sadly it is not easily possible to create HTMLCollection or NodeList on the fly - only with workarounds. And those workarounds would probably outrun the performance improvement of the `_queryElements`. So I guess if we wan't to keep this implementation it would have to be some kind of "extra" if-check anyway at some point. Unless we find another good way how `_queryElements` handles the missing id. We could return an empty array instead of an array with null if the selector cannot be found. This way `_queryElements` would return something iterable with nothing in it:

```js
            if (regexEl) {
                if (document.getElementById(regexEl[1]) === null) {
                    return [];
                }

                return [document.getElementById(regexEl[1])];
            }
```

aragon999 commented:
@tobiasberge yes I was thinking about this solution as well, I came up with:
```
            if (regexEl) {
                const el = document.getElementById(regexEl[1]);

                return (el) ? [el] : [];
            }
```
to avoid calling `document.getElementById()` twice. Not sure if it really matters, I have not checked.

I did not go with this fix, since I was not sure if it could happen in other cases as well and just to be save I decided for this solution. But as you have investigated this, I would be happy to implement this change. Should I immediately do it, or will you revert the original change anyways for 6.4.8.1?

aragon999 commented:
@tobiasberge I now implemented the changes, and added a few more tests for the other cases.

Edit: Sorry I once again updated the PR to move the console test into the `beforeEach` and `afterEach` functions.

tobiasberge commented:
Thanks @aragon999 for taking care of this 👍 

I would also prefer the solution where the `getElementById` is stored in a variable just to be sure. Mine was only a rough idea / example. Thanks also for providing more tests, really appreciate it.

But I'm afraid we do not have enough time prior to 6.4.8.1 to fully test this and really make sure that we do not overlook something else in the hurry. That's why we decided for the revert in the first place, otherwise we would have preferred this solution. 

I hope it's ok if we leave this PR open for now. Then after 6.4.8.1 we can restore the commit from @tinect again and cherry-pick your commit onto it to have the fixed version already while keeping both contributions.

@tinect What do you think about returning an empty array when no element is found?

aragon999 commented:
Alright, after the release I will take care of the PR to be updated and to include the revert of the revert :) 

Edit: I already discussed with @tinect before submitting the PR, I think he also was thinking about something like this. In fact the proposal how to detect the console error messages was his :) But yes, it would be good to get some additional feedback :)

tinect commented:
> @tinect What do you think about returning an empty array when no element is found?

@tobiasberge 
Please note: As deterred with @aragon999 yesterday, there is currently `just` an error thrown. There is nothing broken after this :-)
I'm not sure, if I understand the point.
I like the additionall tests and the check for the id-element to be valid :-)

tobiasberge commented:
Hey @tinect, 

yeah now I have seen that the errors all get catched inside `initializePlugins` and then being displayed via `console.error` which made it look like an actual `throw`. I saw the `throw new Error('Passed element is not an Html element!');` and assumed it could break execution in some cases and have gone the "better safe then sorry" route. I will try to clear up if we can still undo the revert, but I cannot promise it.

Sorry for the circumstances.

tobiasberge commented:
Hi @aragon999 @tinect,

we have undone the revert for 6.4.8.1 and we can continue normally with this PR. So no need for cherry-picking etc. It will be imported for review/testing. Thank you for your patience.
